### PR TITLE
JDK-8318183: C2: VM may crash after hitting node limit

### PR DESCRIPTION
**Pull request by tstuefe**
When processing the node limit, as well as when bailing out of compilation for other reasons, we run `Compiler::record_failure()` which sets the root node of Compile to null.

We then should stop any further work that may use Compile::_root. That does not always happen. The results are crashes or asserts when the node limit check hit right at the wrong spot. One example for a follow-up assertion after the node limit hit:

```
# Internal Error (/shared/projects/openjdk/jdk-jdk/source/src/hotspot/share/opto/loopnode.hpp:623), pid=761308, tid=761322
# assert(_head != nullptr) failed: precond
```
at
```
Stack: [0x00007fa068281000,0x00007fa068382000], sp=0x00007fa06837cd40, free space=1007k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V [libjvm.so+0x114ad1c] IdealLoopTree::IdealLoopTree(PhaseIdealLoop*, Node*, Node*)+0x14c (loopnode.hpp:623)
V [libjvm.so+0x11412bf] PhaseIdealLoop::build_and_optimize()+0x237 (loopnode.cpp:4325)
V [libjvm.so+0x96e7ac] PhaseIdealLoop::PhaseIdealLoop(PhaseIterGVN&, LoopOptsMode)+0x144 (loopnode.hpp:1112)
V [libjvm.so+0x96e9fa] PhaseIdealLoop::optimize(PhaseIterGVN&, LoopOptsMode)+0x4a (loopnode.hpp:1191)
V [libjvm.so+0x960394] Compile::Optimize()+0x98c (compile.cpp:2357)
```

I saw this during the development of [JDK-8318016](https://bugs.openjdk.org/browse/JDK-8318016), which re-uses the node limit check for memory limit checks, so those code paths are executed more often.

The patch adds several bailouts to the C2 code. I tried to follow and close all paths that call the node limit check. With this patch above assertion and other crashes and assertions disappear.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318183](https://bugs.openjdk.org/browse/JDK-8318183): C2: VM may crash after hitting node limit (**Bug** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16205/head:pull/16205` \
`$ git checkout pull/16205`

Update a local copy of the PR: \
`$ git checkout pull/16205` \
`$ git pull https://git.openjdk.org/jdk.git pull/16205/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16205`

View PR using the GUI difftool: \
`$ git pr show -t 16205`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16205.diff">https://git.openjdk.org/jdk/pull/16205.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16205#issuecomment-1766010400)

### CODE REVIEW

File: src/hotspot/share/opto/parse1.cpp [hunk 1]
```diff
@@ -1575,6 +1575,7 @@ void Parse::do_one_block() {
 #endif //ASSERT
 
     do_one_bytecode();
+    if (C->failing()) return;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/parse1.cpp:1578]: I think you can simple call `failing` here and potentially at other places as well.

Comment (tstuefe) [reply src/hotspot/share/opto/parse1.cpp:1578]: Okay. Have not found any other places where that is possible though.

**Comment by tstuefe** [comment]
@TobiHartmann I used GraphKit::failing() as you suggested. One problem I see is that this relies on the local variable C in the Phase super class, and that depends on the phasenumber handed in to the Phase() constructor, and it could theoretically be null. But since all other places that call `failing()` don't seem to care either, I don't think its an issue, right?

**Comment by vnkozlov** [comment]
> @TobiHartmann I used GraphKit::failing() as you suggested. One problem I see is that this relies on the local variable C in the Phase super class, and that depends on the phasenumber handed in to the Phase() constructor, and it could theoretically be null. But since all other places that call `failing()` don't seem to care either, I don't think its an issue, right?

It is only `nullptr` for Compile class where you don't need `C`.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Comment by tstuefe** [comment]
Thank you @vnkozlov and @TobiHartmann !

/integrate

**Pull request closed without merging**